django_ipam/tests/base/test_models.py

```diff
@@ -1,10 +1,89 @@
-class BaseTestIpAddress(object):
-    def test_string_representation(self):
+from ipaddress import IPv4Network, IPv6Network
+
+from django.core.exceptions import ValidationError
+
+
+class BaseTestModel(object):
+    def test_ip_address_string_representation(self):
         ipaddress = self.ipaddress_model(ip_address='entry ip_address')
         self.assertEqual(str(ipaddress), ipaddress.ip_address)
 
+    def test_invalid_ipaddress_subnet(self):
+        failed = True
+        self.subnet_model.objects.create(subnet='192.168.2.0/24')
+        try:
+            self.ipaddress_model.objects.create(ip_address='10.0.0.2',
+                                                subnet=self.subnet_model.objects.first())
+        except ValidationError:
+            failed = False
+        if(failed):
+            self.fail('ValidationError not raised')
+
+    def test_valid_ipaddress_subnet(self):
+        failed = False
+        self.subnet_model.objects.create(subnet='192.168.2.0/24')
+        try:
+            self.ipaddress_model.objects.create(ip_address='192.168.2.1',
+                                                subnet=self.subnet_model.objects.first())
+        except ValidationError:
+            failed = True
+        if(failed):
+            self.fail('ValidationError raised')
+
+    def test_used_ipaddress(self):
+        failed = True
+        self.ipaddress_model.objects.create(ip_address='192.168.1.1')
+        try:
+            self.ipaddress_model.objects.create(ip_address='192.168.1.1')
+        except ValidationError:
+            failed = False
+        if(failed):
+            self.fail('ValidationError not raised')
+
+    def test_subnet_string_representation(self):
+        subnet = self.subnet_model(subnet='entry subnet')
+        self.assertEqual(str(subnet), str(subnet.subnet))
+
+    def test_valid_cidr_field(self):
+        failed = False
+        try:
+            self.subnet_model.objects.create(subnet='22.0.0.0/24')
+        except ValidationError:
+            failed = True
+        if(failed):
+            self.fail('ValidationError raised')
+
+    def test_invalid_cidr_field(self):
+        failed = True
+        try:
+            self.subnet_model.objects.create(subnet='192.192.192.192.192')
+        except ValidationError:
+            failed = False
```

Comment:
check the exception being thrown has something you expect in it, eg: https://github.com/openwisp/openwisp-controller/blob/connections/openwisp_controller/connection/tests/test_models.py#L105